TFTRT: Respect device placement requested by user

meena-at-work opened the pull request:
Previously TRT engines would always default to running on GPU 0. On a multi GPU system, this does not make the best use of the available resources. This PR adds the ability to specify the GPU on which the TRT engine should run.

bixia1 commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -1196,6 +1197,16 @@ def convert(self, calibration_input_fn=None):
+      self._device = array_ops.zeros([]).device
+      if not self._device:  # if device unspecified, defaulting to GPU 0
+        self._device = "gpu:0"
+      if "gpu" not in self._device.lower():
+        raise ValueError(f"Specified device is not a GPU: {self._device}")
+      logging.info(f"Placing imported graph from "
+                   f"`{self._input_saved_model_dir}` on device: {self._device}")
```
Should this be in the same indentation level as line 1200?

meena-at-work replied:
Good point -- doing so will log it for all cases.

meena-at-work replied:
@bixia1  -- done. Please review!

bixia1 commented:
Can we have a test to show how we would handle this situations: some nodes in the imported graph are already assigned with a gpu ?

bixia1 commented:
> @bixia1 -- I was looking into this, and realized that TF2 saved models don't contain device information. Do you mean at test of this nature:
> 
> ```
>     model = create_model()
>     input_saved_model_dir='./saved_model/my_model'
>     model.save(input_saved_model_dir)
> 
>     with tf.device('gpu:1'):
>         model_loaded = tf.saved_model.load(export_dir='./saved_model/my_model')
>         print("Loaded Model device:", model_loaded.variables[0].device)
> 
>     with tf.device('gpu:2'):
>         from tensorflow.python.compiler.tensorrt import trt_convert as trt
>         converter = trt.TrtGraphConverterV2(input_saved_model_dir=input_saved_model_dir)
>         converter.convert()
>         converter.summary()
> ```
> 
> The output for the above with the feature is:
> 
> ```
> INFO:tensorflow:Assets written to: ./saved_model/my_model/assets
> Loaded Model device: /job:localhost/replica:0/task:0/device:GPU:1
> INFO:tensorflow:Linked TensorRT version: (8, 2, 4)
> INFO:tensorflow:Loaded TensorRT version: (8, 2, 4)
> INFO:tensorflow:Placing imported graph from `./saved_model/my_model` on device: /job:localhost/replica:0/task:0/device:GPU:2
> TRTEngineOP Name                 Device        # Nodes # Inputs      # Outputs     Input DTypes       Output Dtypes      Input Shapes       Output Shapes     
> ================================================================================================================================================================
> TRTEngineOp_000_000              device:GPU:2  5       1             1             ['float32']        ['float32']        [[-1, 784]]        [[-1, 10]]        
> 
>         - Const: 2x
>         - MatMul: 2x
>         - Relu: 1x
> 
> ================================================================================================================================================================
> [*] Total number of TensorRT engines: 1
> [*] % of OPs Converted: 71.43% [5/7]
> ```
> 
> Support for TF1 saved models has not been added yet, and is in WIP.

Thanks for explaining this.

meena-at-work commented:
@bixia1 , @DEKHTIARJonathan  -- I've added support for assigning device to TF1 models as well, please take a look.

bixia1 commented:
Put the original PR description here
Previously TRT engines would always default to running on GPU 0. On a multi GPU system, this does not make the best use of the available resources. This PR adds the ability to specify the GPU on which the TRT engine should run.

So now, when the user requests a specific device, TFTRT will convert and run the engine on that specified device. e.g.:

with tf.device('gpu:2'):
    from tensorflow.python.compiler.tensorrt import trt_convert as trt
    converter = trt.TrtGraphConverterV2(input_saved_model_dir=input_saved_model_dir)
    converter.convert()
the engine will be built and run on GPU 2, if it exists on the system.

Running converter.summary() will report the actual GPU on which the engine was placed, e.g. for the above case:

Found the following GPUs:
        - PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')
        - PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU')
        - PhysicalDevice(name='/physical_device:GPU:2', device_type='GPU')
INFO:tensorflow:Assets written to: ./saved_model/my_model/assets
INFO:tensorflow:Linked TensorRT version: (8, 2, 1)
INFO:tensorflow:Loaded TensorRT version: (8, 2, 3)
INFO:tensorflow:Loaded TensorRT 8.2.3 and linked TensorFlow against TensorRT 8.2.1. This is supported because TensorRT minor/patch upgrades are backward compatible.
INFO:tensorflow:Placing imported graph from `./saved_model/my_model` on device: /job:localhost/replica:0/task:0/device:GPU:2
TRTEngineOP Name                 Device        # Nodes # Inputs      # Outputs     Input DTypes       Output Dtypes      Input Shapes       Output Shapes     
================================================================================================================================================================
TRTEngineOp_0_0                  device:GPU:2  5       1             1             ['float32']        ['float32']        [[-1, 784]]        [[-1, 10]]        

        - Const: 2x
        - MatMul: 2x
        - Relu: 1x

================================================================================================================================================================
[*] Total number of TensorRT engines: 1
[*] % of OPs Converted: 71.43% [5/7]

meena-at-work committed:
TFTRT: Respect device placement requested by user

Previously TRT engines would always default to
running on GPU 0. On a multi GPU system, this
does not make the best use of the available
resources. This commit adds the ability to specify
the GPU on which the TRT engine should run.

Signed-off-by: Meenakshi Venkataraman <meenakshiv@nvidia.com>

meena-at-work commented:
A howto for the feature. The user can request for engine building and inference on a specific device, and TFTRT will convert and run the engine on that specified device. e.g.:

```
with tf.device('gpu:2'):
    from tensorflow.python.compiler.tensorrt import trt_convert as trt
    converter = trt.TrtGraphConverterV2(input_saved_model_dir=input_saved_model_dir)
    converter.convert()
```

the engine will be built and run on  GPU 2, if it exists on the system.

Running `converter.summary()` will report the actual GPU on which the engine was placed, e.g. for the above case:

```
Found the following GPUs:
        - PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')
        - PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU')
        - PhysicalDevice(name='/physical_device:GPU:2', device_type='GPU')
INFO:tensorflow:Assets written to: ./saved_model/my_model/assets
INFO:tensorflow:Linked TensorRT version: (8, 2, 1)
INFO:tensorflow:Loaded TensorRT version: (8, 2, 3)
INFO:tensorflow:Loaded TensorRT 8.2.3 and linked TensorFlow against TensorRT 8.2.1. This is supported because TensorRT minor/patch upgrades are backward compatible.
INFO:tensorflow:Placing imported graph from `./saved_model/my_model` on device: /job:localhost/replica:0/task:0/device:GPU:2
TRTEngineOP Name                 Device        # Nodes # Inputs      # Outputs     Input DTypes       Output Dtypes      Input Shapes       Output Shapes     
================================================================================================================================================================
TRTEngineOp_0_0                  device:GPU:2  5       1             1             ['float32']        ['float32']        [[-1, 784]]        [[-1, 10]]        

        - Const: 2x
        - MatMul: 2x
        - Relu: 1x

================================================================================================================================================================
[*] Total number of TensorRT engines: 1
[*] % of OPs Converted: 71.43% [5/7]
```

meena-at-work commented:
@bixia1 -- I've shortened the PR description & added a howto in one of the comments.

Also, this work is now rebased on top of @tfeher 's PR: https://github.com/tensorflow/tensorflow/pull/55166, as they touch the same files.

bixia1 commented:
Many tests failed, will email a log

meena-at-work commented:
@bixia1  -- fixed the bug, please retest. Thanks!
